mllib/src/test/scala/org/apache/spark/ml/image/ImageSchemaSuite.scala

```diff
@@ -0,0 +1,124 @@
+    val rows = Seq(Row(Row(origin, height, width, nChannels, mode, data)),
+      Row(Row(null, height, width, nChannels, mode, data)))
+    val rdd = sc.makeRDD(rows)
+    val df = spark.createDataFrame(rdd, imageDFSchema)
+
+    assert(df.count == 2, "incorrect image count")
+    assert(ImageSchema.isImageColumn(df, "image"), "data do not fit ImageSchema")
+  }
+
+  test("readImages count test") {
+    var df = readImages(imagePath, recursive = false)
+    assert(df.count == 0)
+
+    df = readImages(imagePath, recursive = true, dropImageFailures = false)
+    assert(df.count == 8)
+
+    df = readImages(imagePath, recursive = true, dropImageFailures = true)
+    val count100 = df.count
+    assert(count100 == 7)
+
+    df = readImages(imagePath, recursive = true, sampleRatio = 0.5, dropImageFailures = true)
+    // Random number about half of the size of the original dataset
+    val count50 = df.count
+    assert(count50 > 0.2 * count100 && count50 < 0.8 * count100)
+  }
+
+  test("readImages partition test") {
+    val df = readImages(imagePath, recursive = true, dropImageFailures = true, numPartitions = 3)
+    assert(df.rdd.getNumPartitions == 3)
+  }
+
+  // Images with the different number of channels
+  test("readImages pixel values test") {
+
+    val images = readImages(imagePath + "/multi-channel/", recursive = false).collect
+
+    images.foreach{
+      rrow => {
+        val row = rrow.getAs[Row](0)
+        val filename = Paths.get(getOrigin(row)).getFileName().toString()
+        if(firstBytes20.contains(filename)) {
+          val mode = getMode(row)
+          val bytes20 = getData(row).slice(0, 20)
+
+          val expectedMode = firstBytes20(filename)._1
+          val expectedBytes = firstBytes20(filename)._2
+
+          assert(expectedMode == mode, "mode of the image is not read correctly")
+
+          if (!compareBytes(expectedBytes, bytes20)) {
+            throw new Exception("incorrect numeric value for flattened image")
+          }
+        }
+      }
+    }
+  }
+
+  // number of channels and first 20 bytes of OpenCV representation
+  // - default representation for 3-channel RGB images is BGR row-wise:
+  //   (B00, G00, R00,      B10, G10, R10,      ...)
+  // - default representation for 4-channel RGB images is BGRA row-wise:
+  //   (B00, G00, R00, A00, B10, G10, R10, A00, ...)
+  private val firstBytes20 = Map(
+    "grayscale.png" ->
+      (("CV_8UC1", Array[Byte](0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 3, 5, 2, 1))),
+    "RGB.png" -> (("CV_8UC3",
+      Array[Byte](-34, -66, -98, -38, -69, -98, -62, -90, -117,
+        -70, -98, -124, -34, -63, -90, -20, -48, -74, -18, -45))),
+    "RGBA.png" -> (("CV_8UC4",
+      Array[Byte](-128, -128, -8, -1, -128, -128, -8, -1, -128,
+        -128, -8, -1, 127, 127, -9, -1, 127, 127, -9, -1)))
+  )
+
+  private def compareBytes(x: Array[Byte], y: Array[Byte]): Boolean = {
+    val length = Math.min(x.length, y.length)
```

Comment:
Isn't there a JDK method that compares arrays?

Reply:
done, used Arrays.equals()